### ISSUE TITLE
concurrent Test Data Builder support

### ISSUE THREAD

**Issue by xenoterracide** [immutables/immutables#1276]
https://ardalis.com/improve-tests-with-the-builder-pattern-for-test-data/ e.g. maybe stuff like this, and it could output, whatever to the test generated sources by default.

```java
@Value.Parameter( testOnly=true )
Long getId()
```

**Comment by elucash** [comment]
I don't really understand how `testOnly` should work, however it seems that you can just create test instances and then use `Builder.from(testSample)` to fill all the default values before overriding specific ones.

**Comment by xenoterracide** [comment]
the idea I'm trying to hand out is that `testOnly` could create a second builder in either the test source, for things that you wouldn't want to be true in a production instance. like

```java
interface CountrySubdivision {
    /
    @Value.TestOnly( default=true )
    default CountrySubdivisionType code() {
         return CountrySubdivisionType.DE; // Delaware
     }
}
``` 

I would never want a production version to return Deleware by default, it should always call the "builder setter", but in a test, I love the idea of defaults. So test only, or whatever API would generate another builder in the test layer.

I think ideally this would be done by another jar than value, since for gradle you would want to add that dependency to the new-ish `testFixtures` sourceset that sits between the `main` and `test` sourcesets.

**Comment by xenoterracide** [comment]
Another example is I don't want fields to be nullable in production, but in some cases I'd love for them to be able to be nullable in some tests (I'm lazy and don't actually want to create the whole object structure)

**Comment by xenoterracide** [comment]
do you understand now? I mean the biggest thing here is to be able to have default values in a test builder.

For example, in Brix I have my [Application](https://github.com/xenoterracide/brix/blob/2a8bab70e6aee49b2fcc0a17bf2d2528da4481e7/src/main/java/com/xenoterracide/brix/Application.java) class, which implements [CliConfiguraton](https://github.com/xenoterracide/brix/blob/2a8bab70e6aee49b2fcc0a17bf2d2528da4481e7/src/main/java/com/xenoterracide/brix/CliConfiguration.java), so that's my production implementation. For a test however it'd be nice to just be able to do this, without making yet another interface, or abstract class for testing. maybe something like:

```java
@Value.Style( defaultsTestOnly = true )
interface CliConfiguration extends LoggingConfiguration {
  default Path getConfigDir() { ... }

  default Path getWorkdir() { ... }

  default String getProject() { ... }

  default String getLanguage() { ... }

  default String getModuleType() { ... }

  default @Nullable String getName() { ... }
}
```

so the `main` generated builder wouldn't see the "default" it'd act like it wasn't there, but a test implementation would also be generated, and that would use the values provided in the default.

**Comment by xenoterracide** [comment]
@elucash I might have a reasonable way to achieve this, with one problem, the lack of a shared implementation class. Though I'm not certain how much of a problem that is. In one case what I've done is simply extended the base interface in my test, and provide the defaults there, though that is a little annoying. Reading it again, your way might work too, if I wrap that in an object that delegates, which I was likely to do anyways.

**Comment by elucash** [comment]
I've reread this, but still I somehow feel the use case is easily addressed by simpler means. Code generation should not be used to sweep everything under the rug, it provides some frequently used basics (like default, derived, lazy etc) and some other things which cannot be compensated by other means (like guaranteed interning) or maybe for massive scaffolding. The key is that if you can trivially achieve something with just function call or library - the code generation would not be necessary. The advanced stuff like tailoring generated objects to certain way of testing most probably is out of scope and the job for libraries or external/another code generator. The additional complication is mixing compilation and (test) execution concerns. For testing just create sample instances and fill the data from them.

```java
interface TestOnlySamples {
  SomeConfiguration  cliConfiguration = ImmutableSomeConfiguration.builder()
     .moduleType("sample")
     .country(CountrySubdivisionType.DE)
     ...
     .build();
}

// test code then does this
ImmutableSomeConfiguration.builder().from(TestOnlySamples.cliConfiguration)
    .someEssentialField("value")
    .build()

```

As I was writing this, even simpler, dumb but effective way came to mind, just builder factories with pre-filled builders.

```java
class Samples {
    static SomeConfiguration.Builder someConfiguration()  {
       return  ImmutableSomeConfiguration.builder()
         .moduleType("sample")
         .country(CountrySubdivisionType.DE)
    }
}

// in test 
Samples.someConfiguration()
   .someEssentialField("value")
   .build()
```

You can pass builders as parameters to pre-fill them with test data. If you need builders to have common supertype to set or build different instances with the same setup, you can use inner builder definitions, see example with `VehicleBuilder` under https://immutables.github.io/immutable.html#builder 

Yet another, quite draconian way would be to have special static method which checks either static or threadlocal state to understand if it runs in test or not.

```java
public static <T> T testOnly(T value, String attribute) {
   if (weRunningInTest()) return value;
   throw new IllegalStateException( ...  missing value for attribute .. )
}

import static ... testOnly;

// this is almost one to one identical to your first example in terms of key strokes
// and it would never execute in prod, unless attribute is missing and exception is thrown anyway

interface CountrySubdivision {
    @Value.Default
    default CountrySubdivisionType code() {
         return testOnly(CountrySubdivisionType.DE);
     }
}
```

**Comment by xenoterracide** [comment]
Looking back at this again. You're combining a Test Data Builder with Object Mothers, which isn't an antipattern. However, I personally think putting all of the defaults into the object mother might be. Rather, needing an object mother in all cases is an antipattern.

https://reflectoring.io/objectmother-fluent-builder/
https://betterprogramming.pub/why-you-should-use-test-data-builders-714eb9de20c1
https://martinfowler.com/bliki/ObjectMother.html

I've learned the hard way that test data factories like these are an antipattern.
https://eliasnogueira.com/test-data-factory-why-and-how-to-use/

Honestly anything using a static factory for test data is code smell. I found that many times I needed to use injection, and hide the saving to the database in ways that none of these show. This is all to reduce repetitive code in the actual tests.

Just thinking about this again, and how I'm about to handle it this time

The last time I gave a stab at this. I tried to Have an `interface MyEntity` with a `interface MyEntityTD extends MyEntity` (test data builder). Of course there were 2 problems here.

1. Immutables warns about this as unsupported
2. I'd prefer the Builder only return `MyEntity`, although that's not that big of a deal.
